File: leetcode/hard/1857_largestPathValue.py
from typing import List
import json
from collections import deque

class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        n = len(colors)
        childCount = [0] * n
        parents = [[] for i in range(n)]
        colorsByNodes = [[0]*26 for i in range(n)]

        for s, e in edges:
            childCount[s] += 1
            parents[e].append(s)

        nodesQueue = deque()
        for i in range(n):
            nodeColor = ord(colors[i]) - ord('a')
            colorsByNodes[i][nodeColor] = 1

            if childCount[i] == 0:
                nodesQueue.append(i)

        res = 1
        while nodesQueue:
            node = nodesQueue.popleft()
            for p in parents[node]:
                childCount[p] -= 1

                parentColor = ord(colors[p]) - ord('a')

                for c in range(26):
                    countColors = colorsByNodes[node][c] + 1 if parentColor == c else colorsByNodes[node][c]
                    colorsByNodes[p][c] = max(colorsByNodes[p][c], countColors)

                if childCount[p] == 0:
                    nodesQueue.append(p)
                    res = max(res, max(colorsByNodes[p]))



        if sum(childCount) != 0:
            return -1

        return res



def test ():
    params = [
        {
            'input': ["g", []],
            'output': 1,
        },
        {
            'input': ["abaca", [[0,1],[0,2],[2,3],[3,4]]],
            'output': 3,
        },
        {
            'input': ["a", [[0,0]]],
            'output': -1,
        },
    ]
    solution = Solution()

    for param in params:
        colors, edges = param['input']

        result = solution.largestPathValue(colors, edges)
        correct = json.dumps(result) == json.dumps(param['output'])

        msg = 'SUCCESS' if correct else 'ERROR'
        msg += '\n'
        if not correct:
            msg += 'input ' + json.dumps(param['input']) + '\n'
            msg += 'output ' + json.dumps(param['output']) + '\n'
            msg += 'result ' + json.dumps(result) + '\n'

        print(msg)


if __name__ == '__main__':
    test()

File: leetcode/hard/1857_largestPathValue.ts
function largestPathValue(colors: string, edges: number[][]): number {
    const visited: number[] = new Array(colors.length).fill(0);
    const memo: number[][] = new Array(colors.length).fill(0).map(() => new Array(26).fill(0));
    const adj: number[][] = new Array(colors.length).fill(0).map(() => []);
    for (const [s, e] of edges) {
        adj[s].push(e);
    }

    const aCode = 'a'.charCodeAt(0);
    let res: number = 1;

    const dfs = (id: number): void => {
        if (visited[id] === 1 || res === Number.POSITIVE_INFINITY) {
            res = Number.POSITIVE_INFINITY;
            return;
        }

        if (visited[id] === 2) {
            return;
        }

        visited[id] = 1;

        const idColor: number = colors[id].charCodeAt(0) - aCode;
        memo[id][idColor] = 1;

        for (const nei of adj[id]) {
            if (visited[nei] === 1) {
                res = Number.POSITIVE_INFINITY;
                return;
            }

            dfs(nei);

            for (let i: number = 0; i < 26; ++i) {
                let newColorCount: number = memo[nei][i];
                if (idColor === i) {
                    newColorCount += 1;
                }
                memo[id][i] = Math.max(memo[id][i], newColorCount);
            }

            res = Math.max(res, ...memo[id]);
        }

        visited[id] = 2;
    }

    for (let i: number = 0; i < colors.length; ++i) {
        dfs(i);
    }

    return (res === Number.POSITIVE_INFINITY) ? -1 : res;
};

const test = () => {
    const params = [
        {
            input: {
                colors: "abaca", edges: [[0,1],[0,2],[2,3],[3,4]]
            },
            output: 3,
        },
        {
            input: {
                colors: "a", edges: [[0,0]],
            },
            output: -1,
        },
    ];

    params.forEach(({input, output}) => {
        const { colors, edges } = input;
        const result = largestPathValue(colors, edges);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();